# Module 5 Homework - Spark 

## Question 1:
### Install Spark and PySpark

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?

In [1]:
import pyspark
from pyspark.sql import SparkSession

In [2]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/03/03 21:33:52 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [3]:
spark.version

'3.3.2'

## Question 2:
### FHV October 2019

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.

Repartition the Dataframe to 6 partitions and save it to parquet.

What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

- 1MB
- 6MB
- 25MB
- 87MB

In [4]:
# Download the file
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/fhv_tripdata_2019-10.parquet

--2024-03-03 21:34:07--  https://d37ci6vzurychx.cloudfront.net/trip-data/fhv_tripdata_2019-10.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 65.8.245.178, 65.8.245.51, 65.8.245.171, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|65.8.245.178|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18344035 (17M) [application/x-www-form-urlencoded]
Saving to: ‘fhv_tripdata_2019-10.parquet.1’

fhv_tripdata_2019-1 100%[===================>]  17.49M  55.9MB/s    in 0.3s    

2024-03-03 21:34:08 (55.9 MB/s) - ‘fhv_tripdata_2019-10.parquet.1’ saved [18344035/18344035]



In [6]:
df = spark.read \
    .option("header", "true") \
    .parquet('./fhv_tripdata_2019-10.parquet')

In [7]:
df.schema

StructType([StructField('dispatching_base_num', StringType(), True), StructField('pickup_datetime', TimestampType(), True), StructField('dropOff_datetime', TimestampType(), True), StructField('PUlocationID', DoubleType(), True), StructField('DOlocationID', DoubleType(), True), StructField('SR_Flag', IntegerType(), True), StructField('Affiliated_base_number', StringType(), True)])

In [8]:
df_repartitioned = df.repartition(6)

In [10]:
df_repartitioned.write.parquet('data/pq/fhv/', mode='overwrite')

In [11]:
!ls -lh ./data/pq/fhv

total 39M
-rw-r--r-- 1 juangrau juangrau    0 Mar  3 21:35 _SUCCESS
-rw-r--r-- 1 juangrau juangrau 6.4M Mar  3 21:35 part-00000-f76ff9dc-aea1-454a-b926-93405a2d48da-c000.snappy.parquet
-rw-r--r-- 1 juangrau juangrau 6.4M Mar  3 21:35 part-00001-f76ff9dc-aea1-454a-b926-93405a2d48da-c000.snappy.parquet
-rw-r--r-- 1 juangrau juangrau 6.4M Mar  3 21:35 part-00002-f76ff9dc-aea1-454a-b926-93405a2d48da-c000.snappy.parquet
-rw-r--r-- 1 juangrau juangrau 6.4M Mar  3 21:35 part-00003-f76ff9dc-aea1-454a-b926-93405a2d48da-c000.snappy.parquet
-rw-r--r-- 1 juangrau juangrau 6.4M Mar  3 21:35 part-00004-f76ff9dc-aea1-454a-b926-93405a2d48da-c000.snappy.parquet
-rw-r--r-- 1 juangrau juangrau 6.4M Mar  3 21:35 part-00005-f76ff9dc-aea1-454a-b926-93405a2d48da-c000.snappy.parquet


Answer: 6M

## Question 3:
### Count records

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

- 108,164
- 12,856
- 452,470
- 62,610

In [12]:
df_repartitioned.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropOff_datetime',
 'PUlocationID',
 'DOlocationID',
 'SR_Flag',
 'Affiliated_base_number']

In [13]:
from pyspark.sql import functions as F

In [14]:
df_with_pu_date = df_repartitioned \
    .withColumn('pickup_date', F.to_date(df_repartitioned['pickup_datetime']))

In [15]:
df_with_pu_date.filter(df_with_pu_date['pickup_date'] == '2019-10-15').count()

62629

In [37]:
import pandas as pd
from datetime import datetime

In [17]:
tmpdf = pd.read_parquet('./fhv_tripdata_2019-10.parquet')

In [42]:
tmpdf['pu_date'] = tmpdf['pickup_datetime'].dt.day

In [43]:
tmpdf.head()

dispatching_base_num     pickup_datetime    dropOff_datetime  PUlocationID  \
0               B00009 2019-10-01 00:23:00 2019-10-01 00:35:00         264.0   
1               B00013 2019-10-01 00:11:29 2019-10-01 00:13:22         264.0   
2               B00014 2019-10-01 00:11:43 2019-10-01 00:37:20         264.0   
3               B00014 2019-10-01 00:56:29 2019-10-01 00:57:47         264.0   
4               B00014 2019-10-01 00:23:09 2019-10-01 00:28:27         264.0   

   DOlocationID SR_Flag Affiliated_base_number  pu_date  
0         264.0    None                 B00009        1  
1         264.0    None                 B00013        1  
2         264.0    None                 B00014        1  
3         264.0    None                 B00014        1  
4         264.0    None                 B00014        1

In [45]:
len(tmpdf[tmpdf['pu_date'] == 15])

62629

## Question 4:
### Longest trip for each day

What is the length of the longest trip in the dataset in hours?

- 631,152.50 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours

In [51]:
from pyspark.sql import types

In [58]:
def delta_hours(pick_dt, drop_dt):
    time_delta = drop_dt - pick_dt
    return time_delta.total_seconds() / 3600

In [59]:
delta_hours_udf = F.udf(delta_hours, returnType=types.DoubleType())

In [71]:
df_trip_hours = df_with_pu_date \
    .withColumn("trip_hours", (F.col("dropOff_datetime").cast("long") - F.col("pickup_datetime").cast("long")) / 3600) \
    .select('trip_hours', 'pickup_datetime', 'dropOff_datetime')

In [75]:
df_trip_hours \
    .orderBy(df_trip_hours['trip_hours'].desc()) \
    .take(1)

[Row(trip_hours=631152.5, pickup_datetime=datetime.datetime(2019, 10, 11, 18, 0), dropOff_datetime=datetime.datetime(2091, 10, 11, 18, 30))]

Answer: 631152.5

## Question 5:
### User Interface

Spark’s User Interface which shows the application's dashboard runs on which local port?

- 80
- 443
- 4040
- 8080

Answer: 4040

## Question 6:
### Least frequent pickup location zone

Load the zone lookup data into a temp view in Spark
Zone Data

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?

- East Chelsea
- Jamaica Bay
- Union Sq
- Crown Heights North

In [76]:
# get the Zone data
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2024-03-03 22:46:17--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.112.3
Connecting to github.com (github.com)|140.82.112.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240303%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240303T224617Z&X-Amz-Expires=300&X-Amz-Signature=3a9c82abe930e94157beb01dbade77ee63cd5ace0a08bbc8a5563ebfce3f0b0f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2024-03-03 22:46:17--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [80]:
df_zones = spark.read.csv('./taxi_zone_lookup.csv', header=True)

In [81]:
df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [90]:
df_zones \
    .select('LocationID', 'Borough', 'Zone') \
    .where(df_zones['LocationID']==264) \
    .show()

+----------+-------+----+
|LocationID|Borough|Zone|
+----------+-------+----+
|       264|Unknown|  NV|
+----------+-------+----+



In [83]:
df_with_pu_date.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|pickup_date|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+
|              B00310|2019-10-17 16:55:36|2019-10-17 16:59:14|       264.0|       208.0|   null|                B03017| 2019-10-17|
|              B01626|2019-10-05 02:25:14|2019-10-05 02:43:43|       264.0|       197.0|   null|                B01626| 2019-10-05|
|              B00789|2019-10-03 22:43:16|2019-10-03 22:52:21|       264.0|       264.0|   null|                B00789| 2019-10-03|
|              B00900|2019-10-09 10:59:32|2019-10-09 11:15:28|       264.0|       258.0|   null|                B00900| 2019-10-09|
|              B02157|2019-10-18 18:19:11|2019-10-18 18:21:10|       264.0| 

In [84]:
df_joined = df_with_pu_date.join(df_zones, df_with_pu_date.PUlocationID == df_zones.LocationID)

In [85]:
df_joined.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+----------+--------+--------------------+------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|pickup_date|LocationID| Borough|                Zone|service_zone|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+-----------+----------+--------+--------------------+------------+
|              B00310|2019-10-17 16:55:36|2019-10-17 16:59:14|       264.0|       208.0|   null|                B03017| 2019-10-17|       264| Unknown|                  NV|         N/A|
|              B01626|2019-10-05 02:25:14|2019-10-05 02:43:43|       264.0|       197.0|   null|                B01626| 2019-10-05|       264| Unknown|                  NV|         N/A|
|              B00789|2019-10-03 22:43:16|2019-10-03 22:52:21|       2

In [95]:
df_joined.registerTempTable('joined')

/home/juangrau/spark/spark-3.3.2-bin-hadoop3/python/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [109]:
df_zones_grouped = spark.sql("""
SELECT 
    Zone, COUNT(Zone) AS Count
FROM
    joined
GROUP BY
    Zone
ORDER BY
    Count ASC
""")

In [110]:
df_zones_grouped.show()

+--------------------+-----+
|                Zone|Count|
+--------------------+-----+
|         Jamaica Bay|    1|
|Governor's Island...|    2|
| Green-Wood Cemetery|    5|
|       Broad Channel|    8|
|     Highbridge Park|   14|
|        Battery Park|   15|
|Saint Michaels Ce...|   23|
|Breezy Point/Fort...|   25|
|Marine Park/Floyd...|   26|
|        Astoria Park|   29|
|    Inwood Hill Park|   39|
|       Willets Point|   47|
|Forest Park/Highl...|   53|
|  Brooklyn Navy Yard|   57|
|        Crotona Park|   62|
|        Country Club|   77|
|     Freshkills Park|   89|
|       Prospect Park|   98|
|     Columbia Street|  105|
|  South Williamsburg|  110|
+--------------------+-----+
only showing top 20 rows



Answer: Jamaica Bay

In [3]:
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/fhvhv_tripdata_2021-01.parquet

--2024-02-26 21:22:31--  https://d37ci6vzurychx.cloudfront.net/trip-data/fhvhv_tripdata_2021-01.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 65.8.245.51, 65.8.245.50, 65.8.245.178, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|65.8.245.51|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 308924937 (295M) [application/x-www-form-urlencoded]
Saving to: ‘fhvhv_tripdata_2021-01.parquet’

fhvhv_tripdata_2021 100%[===================>] 294.61M  65.8MB/s    in 4.6s    

2024-02-26 21:22:35 (63.7 MB/s) - ‘fhvhv_tripdata_2021-01.parquet’ saved [308924937/308924937]



In [34]:
import pandas as pd

In [4]:
tmpdf = pd.read_parquet('./fhvhv_tripdata_2021-01.parquet')
tmpdf.to_csv('./fhvhv_tripdata_2021-01.csv', index=False)

In [3]:
df = spark.read \
    .option("header", "true") \
    .parquet('./fhvhv_tripdata_2021-01.parquet')

In [4]:
df.show()

+-----------------+--------------------+--------------------+-------------------+-------------------+-------------------+-------------------+------------+------------+----------+---------+-------------------+-----+----+---------+--------------------+-----------+----+----------+-------------------+-----------------+------------------+----------------+--------------+
|hvfhs_license_num|dispatching_base_num|originating_base_num|   request_datetime|  on_scene_datetime|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|trip_miles|trip_time|base_passenger_fare|tolls| bcf|sales_tax|congestion_surcharge|airport_fee|tips|driver_pay|shared_request_flag|shared_match_flag|access_a_ride_flag|wav_request_flag|wav_match_flag|
+-----------------+--------------------+--------------------+-------------------+-------------------+-------------------+-------------------+------------+------------+----------+---------+-------------------+-----+----+---------+--------------------+-----------+--

In [6]:
df.schema

StructType([StructField('hvfhs_license_num', StringType(), True), StructField('dispatching_base_num', StringType(), True), StructField('originating_base_num', StringType(), True), StructField('request_datetime', TimestampType(), True), StructField('on_scene_datetime', TimestampType(), True), StructField('pickup_datetime', TimestampType(), True), StructField('dropoff_datetime', TimestampType(), True), StructField('PULocationID', LongType(), True), StructField('DOLocationID', LongType(), True), StructField('trip_miles', DoubleType(), True), StructField('trip_time', LongType(), True), StructField('base_passenger_fare', DoubleType(), True), StructField('tolls', DoubleType(), True), StructField('bcf', DoubleType(), True), StructField('sales_tax', DoubleType(), True), StructField('congestion_surcharge', DoubleType(), True), StructField('airport_fee', DoubleType(), True), StructField('tips', DoubleType(), True), StructField('driver_pay', DoubleType(), True), StructField('shared_request_flag',

In [5]:
df.head(5)

[Row(hvfhs_license_num='HV0003', dispatching_base_num='B02682', originating_base_num='B02682', request_datetime=datetime.datetime(2021, 1, 1, 0, 28, 9), on_scene_datetime=datetime.datetime(2021, 1, 1, 0, 31, 42), pickup_datetime=datetime.datetime(2021, 1, 1, 0, 33, 44), dropoff_datetime=datetime.datetime(2021, 1, 1, 0, 49, 7), PULocationID=230, DOLocationID=166, trip_miles=5.26, trip_time=923, base_passenger_fare=22.28, tolls=0.0, bcf=0.67, sales_tax=1.98, congestion_surcharge=2.75, airport_fee=None, tips=0.0, driver_pay=14.99, shared_request_flag='N', shared_match_flag='N', access_a_ride_flag=' ', wav_request_flag='N', wav_match_flag='N'),
 Row(hvfhs_license_num='HV0003', dispatching_base_num='B02682', originating_base_num='B02682', request_datetime=datetime.datetime(2021, 1, 1, 0, 45, 56), on_scene_datetime=datetime.datetime(2021, 1, 1, 0, 55, 19), pickup_datetime=datetime.datetime(2021, 1, 1, 0, 55, 19), dropoff_datetime=datetime.datetime(2021, 1, 1, 1, 18, 21), PULocationID=152, DO

In [6]:
df.repartition(24)

DataFrame[hvfhs_license_num: string, dispatching_base_num: string, originating_base_num: string, request_datetime: timestamp, on_scene_datetime: timestamp, pickup_datetime: timestamp, dropoff_datetime: timestamp, PULocationID: bigint, DOLocationID: bigint, trip_miles: double, trip_time: bigint, base_passenger_fare: double, tolls: double, bcf: double, sales_tax: double, congestion_surcharge: double, airport_fee: double, tips: double, driver_pay: double, shared_request_flag: string, shared_match_flag: string, access_a_ride_flag: string, wav_request_flag: string, wav_match_flag: string]

In [7]:
df.write.parquet('./fhvhv/2021/01/')

In [8]:
df2 = spark.read.parquet('./fhvhv/2021/01/')

In [9]:
for element in df2.schema:
    print(element)

StructField('hvfhs_license_num', StringType(), True)
StructField('dispatching_base_num', StringType(), True)
StructField('originating_base_num', StringType(), True)
StructField('request_datetime', TimestampType(), True)
StructField('on_scene_datetime', TimestampType(), True)
StructField('pickup_datetime', TimestampType(), True)
StructField('dropoff_datetime', TimestampType(), True)
StructField('PULocationID', LongType(), True)
StructField('DOLocationID', LongType(), True)
StructField('trip_miles', DoubleType(), True)
StructField('trip_time', LongType(), True)
StructField('base_passenger_fare', DoubleType(), True)
StructField('tolls', DoubleType(), True)
StructField('bcf', DoubleType(), True)
StructField('sales_tax', DoubleType(), True)
StructField('congestion_surcharge', DoubleType(), True)
StructField('airport_fee', DoubleType(), True)
StructField('tips', DoubleType(), True)
StructField('driver_pay', DoubleType(), True)
StructField('shared_request_flag', StringType(), True)
StructFiel

## Example of select

In [14]:
!head -3 fhvhv_tripdata_2021-01.csv

hvfhs_license_num,dispatching_base_num,originating_base_num,request_datetime,on_scene_datetime,pickup_datetime,dropoff_datetime,PULocationID,DOLocationID,trip_miles,trip_time,base_passenger_fare,tolls,bcf,sales_tax,congestion_surcharge,airport_fee,tips,driver_pay,shared_request_flag,shared_match_flag,access_a_ride_flag,wav_request_flag,wav_match_flag
HV0003,B02682,B02682,2021-01-01 00:28:09,2021-01-01 00:31:42,2021-01-01 00:33:44,2021-01-01 00:49:07,230,166,5.26,923,22.28,0.0,0.67,1.98,2.75,,0.0,14.99,N,N, ,N,N
HV0003,B02682,B02682,2021-01-01 00:45:56,2021-01-01 00:55:19,2021-01-01 00:55:19,2021-01-01 01:18:21,152,167,3.65,1382,18.36,0.0,0.55,1.63,0.0,,0.0,17.06,N,N, ,N,N


In [10]:
# This command is lazy, this means it doesn't change anything until a show is requested
df2.select('pickup_datetime', 'dropoff_datetime', 'PULocationID', 'DOLocationID') \
    .filter(df2['hvfhs_license_num'] == 'HV0003') \
    .show()

+-------------------+-------------------+------------+------------+
|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|
+-------------------+-------------------+------------+------------+
|2021-01-01 00:33:44|2021-01-01 00:49:07|         230|         166|
|2021-01-01 00:55:19|2021-01-01 01:18:21|         152|         167|
|2021-01-01 00:23:56|2021-01-01 00:38:05|         233|         142|
|2021-01-01 00:42:51|2021-01-01 00:45:50|         142|         143|
|2021-01-01 00:48:14|2021-01-01 01:08:42|         143|          78|
|2021-01-01 00:14:30|2021-01-01 00:50:27|          71|         226|
|2021-01-01 00:22:54|2021-01-01 00:30:20|         112|         255|
|2021-01-01 00:40:12|2021-01-01 00:53:31|         255|         232|
|2021-01-01 00:56:45|2021-01-01 01:17:42|         232|         198|
|2021-01-01 00:29:04|2021-01-01 00:36:27|         113|          48|
|2021-01-01 00:48:56|2021-01-01 00:59:12|         239|          75|
|2021-01-01 00:11:53|2021-01-01 00:18:06|       

## sql functions with pyspark

In [23]:
from pyspark.sql import functions as F
from pyspark.sql import types

In [12]:
df2 \
    .withColumn('pickup_date', F.to_date(df2['pickup_datetime'])) \
    .withColumn('dropoff_date', F.to_date(df2['dropoff_datetime'])) \
    .select('pickup_date', 'dropoff_date', 'PULocationID', 'DOLocationID') \
    .show()

+-----------+------------+------------+------------+
|pickup_date|dropoff_date|PULocationID|DOLocationID|
+-----------+------------+------------+------------+
| 2021-01-01|  2021-01-01|         230|         166|
| 2021-01-01|  2021-01-01|         152|         167|
| 2021-01-01|  2021-01-01|         233|         142|
| 2021-01-01|  2021-01-01|         142|         143|
| 2021-01-01|  2021-01-01|         143|          78|
| 2021-01-01|  2021-01-01|          88|          42|
| 2021-01-01|  2021-01-01|          42|         151|
| 2021-01-01|  2021-01-01|          71|         226|
| 2021-01-01|  2021-01-01|         112|         255|
| 2021-01-01|  2021-01-01|         255|         232|
| 2021-01-01|  2021-01-01|         232|         198|
| 2021-01-01|  2021-01-01|         113|          48|
| 2021-01-01|  2021-01-01|         239|          75|
| 2021-01-01|  2021-01-01|         181|         237|
| 2021-01-01|  2021-01-01|         236|          68|
| 2021-01-01|  2021-01-01|         256|       

In [62]:
# defining an function
def crazy_stuff(base_num):
    '''
    num = int(base_num[1:])
    if num % 7 == 0:
        return f's/{num:03x}'
    elif num % 3 == 0:
        return f'a/{num:03x}'
    else:
        return f'e/{num:03x}'
    '''
    return 'hello'

In [63]:
crazy_stuff('B01100')

'hello'

In [64]:
crazy_stuff_udf = F.udf(crazy_stuff, returnType=types.StringType())

In [66]:
df2 \
    .withColumn('pickup_date', F.to_date(df2['pickup_datetime'])) \
    .withColumn('dropoff_date', F.to_date(df2['dropoff_datetime'])) \
    .withColumn('base_num', crazy_stuff_udf(df2['dispatching_base_num'])) \
    .select('pickup_date', 'dropoff_date', 'PULocationID', 'DOLocationID') \
    .show()

+-----------+------------+------------+------------+
|pickup_date|dropoff_date|PULocationID|DOLocationID|
+-----------+------------+------------+------------+
| 2021-01-01|  2021-01-01|         230|         166|
| 2021-01-01|  2021-01-01|         152|         167|
| 2021-01-01|  2021-01-01|         233|         142|
| 2021-01-01|  2021-01-01|         142|         143|
| 2021-01-01|  2021-01-01|         143|          78|
| 2021-01-01|  2021-01-01|          88|          42|
| 2021-01-01|  2021-01-01|          42|         151|
| 2021-01-01|  2021-01-01|          71|         226|
| 2021-01-01|  2021-01-01|         112|         255|
| 2021-01-01|  2021-01-01|         255|         232|
| 2021-01-01|  2021-01-01|         232|         198|
| 2021-01-01|  2021-01-01|         113|          48|
| 2021-01-01|  2021-01-01|         239|          75|
| 2021-01-01|  2021-01-01|         181|         237|
| 2021-01-01|  2021-01-01|         236|          68|
| 2021-01-01|  2021-01-01|         256|       